opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -6,82 +6,220 @@
+    this.mappingType = mappingType;
+  }
+
   /**
-   * JDBC type name.
+   * A constructor function which builds proper `OpenSearchDataType` for given {@link ExprType}.
+   * @param type A type.
+   * @return An instance of `OpenSearchDataType`.
    */
-  private final String jdbcType;
+  public static OpenSearchDataType of(ExprType type) {
+    if (type instanceof OpenSearchDataType) {
+      return (OpenSearchDataType) type;
+    }
+    return new OpenSearchDataType((ExprCoreType) type);
+  }
 
-  @Override
-  public List<ExprType> getParent() {
-    return parents;
+  protected OpenSearchDataType(ExprCoreType type) {
+    this.exprCoreType = type;
   }
 
+  protected OpenSearchDataType() {
+  }
+
+  // object and nested types have properties - info about embedded types
+  @Getter
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> properties = new LinkedHashMap<>();
+
+  // text could have fields
+  @Getter
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> fields = new LinkedHashMap<>();
+
   @Override
   public String typeName() {
-    return jdbcType;
+    if (mappingType == null) {
+      return exprCoreType.typeName();
+    }
+    return mappingType.toString().toLowerCase();
   }
 
   @Override
   public String legacyTypeName() {
-    return LEGACY_TYPE_NAME_MAPPING.getOrDefault(this, typeName());
+    return typeName();
+  }
+
+  /**
+   * Clone type object without {@link #properties} - without info about nested object types.
+   * @return A cloned object.
+   */
+  protected OpenSearchDataType cloneEmpty() {
+    var copy = mappingType != null ? of(mappingType) : new OpenSearchDataType(exprCoreType);
+    copy.fields = fields; //TODO do we need to clone object?
+    copy.exprCoreType = exprCoreType;
+    return copy;
+  }
+
+  /**
+   * Flattens mapping tree into a single layer list of objects (pairs of name-types actually),
+   * which don't have nested types.
+   * See {@link OpenSearchDataTypeTest#traverseAndFlatten() test} for example.
+   * @param tree A list of `OpenSearchDataType`s - map between field name and its type.
+   * @return A list of all `OpenSearchDataType`s from given map on the same nesting level (1).
+   *         Nested object names are prefixed by names of their host.
+   */
+  public static Map<String, OpenSearchDataType> traverseAndFlatten(
+      Map<String, OpenSearchDataType> tree) {
+    Map<String, OpenSearchDataType> result = new LinkedHashMap<>();
+    for (var entry : tree.entrySet()) {
+      result.put(entry.getKey(), entry.getValue().cloneEmpty());
+      result.putAll(
+          traverseAndFlatten(entry.getValue().properties)
+              .entrySet().stream()
+              .collect(
+                  LinkedHashMap::new,
+                  (map, item) -> map.put(
+                      String.format("%s.%s", entry.getKey(), item.getKey()),
+                      item.getValue()),
+                  Map::putAll));
+    }
+    return result;
+  }
```

Comment:
Here's a simpler implementation that also avoid creating two hash maps for each call iterated entry.

```suggestion
  public static Map<String, OpenSearchDataType> traverseAndFlatten(
      Map<String, OpenSearchDataType> tree) {
    final Map<String, OpenSearchDataType> result = new HashMap<>();

    BiConsumer<Map<String, OpenSearchDataType>, String> visitLevel = new BiConsumer<>() {
      @Override
      public void accept(Map<String, OpenSearchDataType> subtree, String prefix) {
        for (var entry : subtree.entrySet()) {
          String entryKey = entry.getKey();
          var nextPrefix = String.format("%s.%s", entryKey, prefix);
          result.put(nextPrefix, entry.getValue().cloneEmpty());

          var nextSubtree = entry.getValue().getProperties();

          if (!nextSubtree.isEmpty()) {
            accept(nextSubtree, nextPrefix);
          }
        }
      }
    };
    visitLevel.accept(tree, "");

    return result;
  }
```

Reply:
Fixed in 902929e1a.